mlflow/_promptlab.py

```diff
@@ -0,0 +1,144 @@
+        results = []
+        for idx in inputs.index:
+            python_inputs = {
+                param.key: inputs["{param.key}"][idx] for param in self.prompt_parameters
+            }
+            prompt = self.prompt_template.substitute(python_inputs)
+            result = mlflow.gateway.query(
+                route=self.model_route,
+                data={
+                    {
+                        "prompt": prompt,
+                    }.update(self.python_parameters),
+                },
+            )
+            results.append(result["candidates"][0]["text"])
+        return results
+
+
+def save_model(
+    promptlab_model,
+    path,
+    conda_env=None,
+    code_paths=None,
+    mlflow_model=None,
+    signature=None,
+    input_example=None,
+    pip_requirements=None,
+    extra_pip_requirements=None,
+):
+    _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
+
+    _validate_and_prepare_target_save_path(path)
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+
+    if mlflow_model is None:
+        mlflow_model = Model()
+    if signature is not None:
+        mlflow_model.signature = signature
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, path)
+
+    model_data_subpath = "model.pkl"
+    model_data_path = os.path.join(path, model_data_subpath)
+    _save_model(promptlab_model, model_data_path)
+
+    pyfunc.add_to_model(
+        mlflow_model,
+        loader_module="_promptlab",
+        model_path=model_data_subpath,
+        conda_env=_CONDA_ENV_FILE_NAME,
+        python_env=_PYTHON_ENV_FILE_NAME,
+        code=code_dir_subpath,
+    )
+
+    mlflow_model.save(os.path.join(path, MLMODEL_FILE_NAME))
+
+    if conda_env is None:
+        if pip_requirements is None:
+            default_reqs = get_default_pip_requirements(True)
+            inferred_reqs = mlflow.models.infer_pip_requirements(
+                path, "_promptlab", fallback=default_reqs
+            )
+            default_reqs = sorted(set(inferred_reqs).union(default_reqs))
+        else:
+            default_reqs = None
+        conda_env, pip_requirements, pip_constraints = _process_pip_requirements(
+            default_reqs, pip_requirements, extra_pip_requirements
+        )
+    else:
+        conda_env, pip_requirements, pip_constraints = _process_conda_env(conda_env)
+
+    with open(os.path.join(path, _CONDA_ENV_FILE_NAME), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    if pip_constraints:
+        write_to(os.path.join(path, _CONSTRAINTS_FILE_NAME), "\n".join(pip_constraints))
+
+    write_to(os.path.join(path, _REQUIREMENTS_FILE_NAME), "\n".join(pip_requirements))
+
+    _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))
+
+
+def _save_model(model, path):
+    with open(path, "wb") as out:
+        import cloudpickle
+
+        cloudpickle.dump(model, out)
```

Comment:
I don't think we should be pickling an instance of `_PromptlabModel`. Instead, `save_model()` should take in   arguments like `prompt`, `temperature`, `max_tokens`, etc., and we should construct an instance of a pyfunc-compatible model wrapper (the same way your loader module code from Scala does) at load time